smithy-model/src/main/java/software/amazon/smithy/model/transform/DeconflictErrorsWithSharedStatusCode.java

```diff
@@ -0,0 +1,158 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package software.amazon.smithy.model.transform;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import software.amazon.smithy.model.Model;
+import software.amazon.smithy.model.knowledge.TopDownIndex;
+import software.amazon.smithy.model.shapes.MemberShape;
+import software.amazon.smithy.model.shapes.OperationShape;
+import software.amazon.smithy.model.shapes.ServiceShape;
+import software.amazon.smithy.model.shapes.Shape;
+import software.amazon.smithy.model.shapes.ShapeId;
+import software.amazon.smithy.model.shapes.StructureShape;
+import software.amazon.smithy.model.shapes.UnionShape;
+import software.amazon.smithy.model.traits.ErrorTrait;
+import software.amazon.smithy.model.traits.HttpErrorTrait;
+import software.amazon.smithy.model.traits.HttpHeaderTrait;
+import software.amazon.smithy.model.traits.Trait;
+import software.amazon.smithy.utils.Pair;
+
+/**
+ * Deconflicts errors on operations that share the same status code by replacing
+ * the conflicting errors with a synthetic error structure that contains hoisted
+ * members that were bound to HTTP headers. The conflicting errors are added to
+ * a union on the synthetic error.
+ */
+public class DeconflictErrorsWithSharedStatusCode {
+
+    private final ServiceShape forService;
+
+    DeconflictErrorsWithSharedStatusCode(ServiceShape forService) {
+        this.forService = forService;
+    }
+
+    Model transform(ModelTransformer transformer, Model model) {
+        // Copy any service errors to the operations to find all potential conflicts.
+        model = transformer.copyServiceErrorsToOperations(model, forService);
+        TopDownIndex topDownIndex = TopDownIndex.of(model);
+        List<Shape> shapesToReplace = new ArrayList<>();
+
+        for (OperationShape operation : topDownIndex.getContainedOperations(forService)) {
+            OperationShape.Builder replacementOperation = operation.toBuilder();
+            boolean replaceOperation = false;
+
+            // Collect errors that share the same status code.
+            Map<String, List<StructureShape>> statusCodesToErrors = new HashMap<>();
+            for (ShapeId errorId : operation.getErrors()) {
+                StructureShape error = model.expectShape(errorId, StructureShape.class);
+                String statusCode = error.hasTrait(HttpErrorTrait.ID)
+                        ? Integer.toString(error.getTrait(HttpErrorTrait.class).get().getCode())
```

Comment:
Why convert the status code to a string? It seems like you could use the Integer value directly in the rest of this transform?